include/secp256k1_musig.h

```diff
@@ -0,0 +1,417 @@
+ *
+ * The workflow for this structure is as follows:
+ *
+ * 1. This structure is initialized with `secp256k1_musig_pubkey_combine`, which sets
+ *    the `pubkey` and `index` fields, and zeros out all other fields.
+ *
+ * 2. If the signer data is used in a non-public session, it should be updated with
+ *    `secp256k1_musig_get_public_nonce` which in addition to returning the nonce
+ *    takes an array of signer data structs and an array of commitments, and sets
+ *    each signer data struct's `nonce_commitment` field to the corresponding
+ *    commitment. If the signer data is used in a public session, the
+ *    `nonce_commitment` is set set in musig_session_initialize_public.
+ *
+ * 3. Each individual data struct should be updated with `secp256k1_musig_set_nonce`
+ *    once a nonce is available. This function takes a single signer data struct
+ *    rather than an array because it may fail in the case that the provided nonce
+ *    does not match the commitment. In this case, it is desirable to identify the
+ *    exact party whose nonce was inconsistent.
+ *
+ * Fields:
+ *   present: indicates whether the signers nonce is set
+ *     index: index of the signer in the MuSig; set to be consistent with the order
+ *            of the pubkeys provided to `secp256k1_musig_pubkey_combine`
+ *    pubkey: public key that the signer will use for partial signing
+ *     nonce: public nonce, must be a valid curvepoint if the signer is `present`
+ * nonce_commitment: commitment to the nonce, or all-bits zero if a commitment
+ *                   has not yet been set
+ */
+typedef struct {
+    int present;
+    size_t index;
+    secp256k1_pubkey pubkey;
+    secp256k1_pubkey nonce;
+    unsigned char nonce_commitment[32];
+} secp256k1_musig_signer_data;
+
+/** Opaque data structure that holds a MuSig partial signature.
+ *
+ *  The exact representation of data inside is implementation defined and not
+ *  guaranteed to be portable between different platforms or versions. It is however
+ *  guaranteed to be 32 bytes in size, and can be safely copied/moved. If you need
+ *  to convert to a format suitable for storage, transmission, or comparison, use the
+ *  `secp256k1_musig_partial_signature_serialize` and
+ *  `secp256k1_musig_partial_signature_parse` functions.
+ */
+typedef struct {
+    unsigned char data[32];
+} secp256k1_musig_partial_signature;
+
+/** Computes a combined public key, hash of keys for use with MuSig, and initializes
+ *  an array of signer data structures
+ *
+ *  Returns: 1 when the public keys were successfully combined, 0 otherwise
+ *  Args:        ctx: pointer to a context object initialized for verification
+ *                    (cannot be NULL)
+ *           scratch: scratch space used to compute the combined pubkey by
+ *                    multiexponentiation. If NULL, an inefficient algorithm is used.
+ *  Out: combined_pk: the MuSig-combined public key (cannot be NULL)
```

Comment:
What do you think about allowing this to be NULL, in case the public key has already been computed (e.g. in a previous signing session)? In that case we would avoid doing the expensive ecmult_multi.

Reply:
Hm, right, if you want to initialize new signer_data but know the pubkey already you still have to call `secp256k1_musig_pubkey_combine` - that's confusing. I'm thinking about renaming that or somehow integrating it with `session_initialize`.

Reply:
For threshold signatures we need a new version of `session_initialize` but not a new version of `pubkey_combine`, so I like the current separation.

Maybe we should just tell users who want to avoid repeated computations that they have to store both the combined pubkey and the freshly-initialized `signer_data` structures?

The workflow I'm considering for threshold signatures is
1. Users call `secp256k1_musig_pubkey_combine` to compute the total MuSig key and initialize `signer_data` structures with each users' original public keys.
2. Each user calls `secp256k1_thresholdsig_keysplit` to split their key into shards, which they send to each user (and also broadcast "public coefficients" which act as a witness that the sharding was done correctly).
3. Recipients of the shards call `secp256k1_thresholdsig_verify_shard` which (a) verifies the public coefficients are consistent with the shards; (b) verifies that the public coefficients are consistent with the original public keys; (c) **updates in place** the `signer_data` structures since after the key-splitting ceremony everybody winds up with modified keys.

So both steps 1 and 3 are kinda expensive and are things that users might want to cache. It would be simple/consistent to tell them "store the `signer_data` structures", though I'm also a bit wary of inviting users to de/serialize this structure because there might be replay attacks.

Reply:
Actually I think it's just super weird to initialize the signers in combined_pk, I hadn't realized that. It should happen in session_init functions.

Reply:
All the steps I just described happen at key-generation time BTW, and in principle only need to happen once.

Reply:
I moved the signer initialization from pubkey_combine to session_initialize.